trading_calendars/exchange_calendar_xist.py

```diff
@@ -0,0 +1,208 @@
+
+    # Note: Effective as of Nov 14, 2016 according to official website
+    Open Time: 10:00 AM, TRT
+    Close Time: 6:00 PM, TRT
+
+    Regularly-Observed Holidays:
+    - New Year's Day
+    - National Sovereignty and Children's Day (Apr 23)
+    - Labour Day (May 1)
+    - Commemoration of Attaturk, Youth and Sports Day (May 19)
+    - Eid-ul-Fitr (Islamic Lunar Calendar)
+    - Democracy and National Unity Day (Jul 15)
+    - Eid-al-Adha (first day is a half-day, remaining full holidays)
+    - Victory Day (Aug 30)
+    - Republic Day (Oct 29)
+
+    Early Closes:
+    - First Day of Eid-al-Adha
+    - Republic Day (Oct 28)
+    """
+    name = 'XIST'
+
+    tz = timezone('Europe/Istanbul')
+
+    open_times = (
+        (None, time(10, 1)),
+    )
+
+    close_times = (
+        (None, time(18, 00)),
+    )
+
+    regular_early_close = time(12, 30)
+
+    @property
+    def regular_holidays(self):
+        return HolidayCalendar([
+            NewYearsDay,
+            NationalSovereigntyAndChildrensDay,
+            LabourDay,
+            CommemorationOfAttaturkYouthAndSportsDay,
+            DemocracyAndNationalUnityDay,
+            VictoryDay,
+            RepublicDay,
+        ])
+
+    @property
+    def adhoc_holidays(self):
+        misc_adhoc = [
+            pd.Timestamp('2002-01-04', tz=UTC),  # Market Holiday
+            pd.Timestamp('2003-02-14', tz=UTC),  # Eid al Adha extra holiday
+            pd.Timestamp('2003-11-21', tz=UTC),  # Terror attacks
+            pd.Timestamp('2003-11-24', tz=UTC),  # Eid al Fitr extra holiday
+            pd.Timestamp('2003-11-28', tz=UTC),  # Eid al Fitr extra holiday
+            pd.Timestamp('2004-01-23', tz=UTC),  # Bad weather
+            pd.Timestamp('2004-12-30', tz=UTC),  # Closure for redenomination
+            pd.Timestamp('2004-12-31', tz=UTC),  # Closure for redenomination
+            pd.Timestamp('2006-01-13', tz=UTC),  # Eid al Adha extra holiday
+        ]
+
+        return list(chain(
+            misc_adhoc,
+            EidAlFitr1,
+            EidAlFitr2,
+            EidAlFitr3,
+            EidAlAdha1,
+            EidAlAdha2,
+            EidAlAdha3,
+            EidAlAdha4,
+        ))
+
+    @property
+    def special_closes(self):
+        return [
+            (
+                self.regular_early_close,
+                HolidayCalendar([RepublicDayHalfDay]),
+            )
+        ]
+
+    @property
+    def special_closes_adhoc(self):
+        # Some early close days fall on holidays, so we must filter those out
+        # of the early close list
+        collisions = [
+            # CAYS Day on May 19
+            pd.Timestamp('1994-05-19'),
+            # Day before Eid al Fitr observed as holiday in 2003
+            pd.Timestamp('2003-11-24'),
+        ]
+
+        early_close_days = EidAlFitrHalfDay + EidAlAdhaHalfDay
+
+        early_close_days = [
+            day for day in early_close_days if day not in collisions
+        ]
```

Comment:
What you have here is good, but just in case you find it useful another way to write this would be:

```python
collisions = set([
    # CAYS Day on May 19
    pd.Timestamp('1994-05-19'),
    # Day before Eid al Fitr observed as holiday in 2003
    pd.Timestamp('2003-11-24'),
])

early_close_days = list(set(EidAlFitrHalfDay + EidAlAdhaHalfDay) - collisions)
```

It's shorter, but yours is clearer I'd say.